gson/src/test/java/com/google/gson/functional/Java17RecordTest.java

```diff
@@ -67,21 +175,201 @@ public void testPrimitiveDefaultValues() {
-    // TODO(eamonnmcmanus): consider forbidding null for primitives
-    String s = "{'aString': 's', 'aByte': null, 'aShort': null, 'anInt': null, 'aLong': null, 'aFloat': null, 'aDouble': null, 'aChar': null, 'aBoolean': null}";
-    assertEquals(expected, gson.fromJson(s, RecordWithPrimitives.class));
+  public void testPrimitiveJsonNullValue() {
+    String s = "{'aString': 's', 'aByte': null, 'aShort': 0}";
+    var e = assertThrows(JsonParseException.class, () -> gson.fromJson(s, RecordWithPrimitives.class));
+    assertEquals("null is not allowed as value for record component 'aByte' of primitive type; at path $.aByte",
+        e.getMessage());
   }
 
-  public record MyRecord(
-      @SerializedName("name") String a,
-      @SerializedName(value = "name1", alternate = {"name2", "name3"}) String b) {
-    public MyRecord {
-      a = "modified-" + a;
-    }
+  /**
+   * Tests behavior when JSON contains non-null value, but custom adapter returns null
+   * for primitive component
+   */
+  @Test
+  public void testPrimitiveAdapterNullValue() {
+    Gson gson = new GsonBuilder()
+        .registerTypeAdapter(byte.class, new TypeAdapter<Byte>() {
+          @Override public Byte read(JsonReader in) throws IOException {
+            in.skipValue();
+            // Always return null
+            return null;
+          }
+
+          @Override public void write(JsonWriter out, Byte value) {
+            throw new AssertionError("not needed for test");
+          }
+        })
+        .create();
+
+    String s = "{'aString': 's', 'aByte': 0}";
+    var exception = assertThrows(JsonParseException.class, () -> gson.fromJson(s, RecordWithPrimitives.class));
+    assertEquals("null is not allowed as value for record component 'aByte' of primitive type; at path $.aByte",
+        exception.getMessage());
   }
 
-  public record RecordWithPrimitives(
+  private record RecordWithPrimitives(
       String aString, byte aByte, short aShort, int anInt, long aLong, float aFloat, double aDouble, char aChar, boolean aBoolean) {}
+
+  /** Tests behavior when value of Object component is missing; should default to null */
+  @Test
+  public void testObjectDefaultValue() {
+    record LocalRecord(String s, int i) {}
+
+    assertEquals(new LocalRecord(null, 1), gson.fromJson("{\"i\":1}", LocalRecord.class));
+  }
+
+  /**
+   * Tests serialization of a record with {@code static} field.
+   *
+   * <p>Important: It is not documented that this is officially supported; this
+   * test just checks the current behavior.
+   */
+  @Test
+  public void testStaticFieldSerialization() {
+    Gson gson = new GsonBuilder()
+        // Include static fields
+        .excludeFieldsWithModifiers(0)
+        .create();
+
+    String json = gson.toJson(new RecordWithStaticField());
+    assertEquals("{\"s\":\"initial\"}", json);
+  }
+
+  /**
+   * Tests deserialization of a record with {@code static} field.
```

Comment:
Is this consistent with existing reflective deserialization? Maybe I'm misreading the code, but it doesn't look to me as if it treats deserialization differently from static. If someone really wants to serialize static fields (which as I've said I don't think they should) then maybe they want to deserialize them too?

Reply:
No, it is not consistent. For simplicity I changed it to always ignore static fields for record deserialization: [relevant lines](https://github.com/google/gson/pull/2224/files/e1f696c641d793081d16c95b22969f136837f37a#diff-925cdeedc560bb3d51ddc67046b656b95247a2a819b4b4b9dd6b4d139526b465R242-R246)

It would be possible to also support deserialization but that would require some adjustments due to record deserialization normally only invoking the canonical constructor. The current `master` state would just throw an exception because the static field has no corresponding canonical constructor parameter.

Though personally I don't think we should allow static field deserialization. It can cause thread safety issues and cause integrity issues when a `static final` field is modified whose constant value was inlined by the compiler at other locations. For records we can still prohibit (respectively ignore) static field deserialization without breaking backward compatibility because record deserialization was mostly not possible before (#1794).

Though maybe this test here does not make the current behavior clear enough?

Reply:
OK, the point about the constructor is an important one. I still hate the idea of serializing _or_ deserializing static fields [:-)] but I can live with serializing them only following explicit configuration, and never deserializing them. Writing static fields is way worse than just reading them.

Reply:
> and never deserializing them

But only for records, right? This pull request keeps the behavior for the regular reflective adapter which supports deserialization of static fields.